File: members/U02BWCHJA6Q/Programers/kakao_internship_2021/거리두기_확인하기.js
// https://programmers.co.kr/learn/courses/30/lessons/81302

function solution(places) {
  const answer = [];
  const LENGTH = 5;

  places.forEach(place => {
    const peoples = [];

    for (let y = 0; y < LENGTH; y++) {
      for (let x = 0; x < LENGTH; x++) {
        if (place[y][x] === 'P') {
          peoples.push({ y, x });
        }
      }
    }

    for (let i = 0; i < peoples.length - 1; i++) {
      for (let j = i + 1; j < peoples.length; j++) {
        const { y: ay, x: ax } = peoples[i];
        const { y: by, x: bx } = peoples[j];

        const distance = Math.abs(ay - by) + Math.abs(ax - bx);

        if (distance > 2) {
          continue;
        }

        if (distance <= 1) {
          answer.push(0);

          return;
        }

        let minX, minY, maxX, maxY;

        if (ay < by) {
          minY = ay;
          maxY = by;
        } else {
          minY = by;
          maxY = ay;
        }

        if (ax < bx) {
          minX = ax;
          maxX = bx;
        } else {
          minX = bx;
          maxX = ax;
        }

        for (let y = minY; y <= maxY; y++) {
          for (let x = minX; x <= maxX; x++) {
            if (place[y][x] === 'O') {
              answer.push(0);

              return;
            }
          }
        }
      }
    }
    answer.push(1);
  });

  return answer;
}

File: members/U02CDR60PFS/kakao2021/거리두기_확인하기.cpp
// 나는 for문으로 모든 조건을 고려해주었다. (거리 2 이하인 가로,세로, 대각선)
// bfs로 짜는게 더 깔끔할 것 같다.
#include <string>
#include <vector>
#include <iostream>
using namespace std;
int checkAround(vector<string> place, int r, int c)
{
    int d[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
    for (int i = 0; i < 4; i++)
    {
        for (int j = 1; j <= 2; j++)
        {
            int nr = r + j * d[i][0];
            int nc = c + j * d[i][1];

            if (nr < 0 || nr > 4 || nc < 0 || nc > 4)
                break;

            if (j == 1)
            {
                if (place[nr][nc] == 'P')
                    return 0;
            }
            else
            {
                if (place[nr][nc] == 'P' && place[nr - d[i][0]][nc - d[i][1]] == 'O')
                    return 0;
            }
        }
    }

    for (int i = 4; i < 8; i++)
    {
        int nr = r + d[i][0];
        int nc = c + d[i][1];

        if (nr < 0 || nr > 4 || nc < 0 || nc > 4)
            continue;

        if (place[nr][nc] == 'P' && (place[r + d[i][0]][c] != 'X' || place[r][c + d[i][1]] != 'X'))
            return 0;
    }

    return 1;
}

int check(vector<string> place)
{
    for (int i = 0; i < 5; i++)
    {
        for (int j = 0; j < 5; j++)
        {
            if (place[i][j] == 'P')
            {
                int rst = checkAround(place, i, j);
                if (rst == 0)
                    return 0;
            }
        }
    }
    return 1;
}

vector<int> solution(vector<vector<string>> places)
{
    vector<int> answer;
    for (int i = 0; i < places.size(); i++)
        answer.push_back(check(places[i]));

    return answer;
}